Gutenboarding - Add editor loading placeholder

Addison-Stavlo opened the pull request:
#### Changes proposed in this Pull Request

Add a loading placeholder for the editor in the Gutenboarding flow as per #39246

![create-site-loading-editor](https://user-images.githubusercontent.com/28742426/73893447-ae28aa80-4847-11ea-956e-b5ba9144f0b9.gif)

This is accomplished by having a `create-site` route at the end of the Gutenboard Steps.  The idea is to update this Route to contain the create-site logic in a follow up (as well as make this redirect to another step if somehow loaded when site creation does not make sense). But I think this is in a pretty good place to test and merge, with room for some followups after #39203 and #39050 land.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*  Run this PR, navigate through Gutenboarding, click "Create Site".
*  Verify there is a loading placeholder until the editor starts loading.
*  After creating the site, visiting `/gutenboarding/create-site` should redirect to `/gutenboarding`

Fixes #39246

simison commented:
I think we should either show just the editor loader:
![image](https://user-images.githubusercontent.com/87168/73829296-68e68780-480b-11ea-8a5f-ba8926c85048.png)

...or the spinner, but not both.

I'd suggest we also cover the whole UI (background+header+create button) with white canvas to create better continuation to a loading screen that shows up in when editor starts loading.

Thoughts @Automattic/gutenboarding-design ?

In initial Frankenflow design had this modal to hide the loading step but since it's little more involved for Fankenflow, we'd just like to go with a simple loading animation as the first step:

![image](https://user-images.githubusercontent.com/87168/73829633-f9bd6300-480b-11ea-88f0-01a7f504fc72.png)

sirreal commented on client/landing/gutenboarding/steps.ts:
```diff
@@ -8,6 +8,7 @@ export enum Step {
 	DesignSelection = '/design',
 	PageSelection = '/pages',
 	Signup = '/signup',
+	SiteLoading = '/siteLoading',
```
A "loading" transitional state doesn't seem like something that deserves a route. How about "create site," which captures the intent when we hit this route?

```suggestion
	CreateSite = '/create-site',
```

cc: @razvanpapadopol

sirreal commented on client/landing/gutenboarding/onboarding-block/loading-placeholder/index.tsx:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Internal dependencies
+ */
+import Spinner from 'components/spinner';
```
😱 Calypso import ⚠️ 

If we wanted to use this, I'd say we'd need to extract it to `@automattic/components`. However, I think we can use a different loading view like @simison [mentioned](https://github.com/Automattic/wp-calypso/pull/39266#issuecomment-582321330) and we shouldn't need this.

simison replied:
Also [core spinner](https://github.com/WordPress/gutenberg/tree/0c089f7a481fae56dd493d679bf7b2ea000b21c4/packages/components/src/spinner) is available.

Addison-Stavlo replied:
Ohh Right... ☠️ Calypso Imports! 💀   .  😆 

Il just remove the spinner for now per @simison 's suggestion.

sirreal commented on client/landing/gutenboarding/onboarding-block/edit.tsx:
```diff
@@ -70,6 +71,11 @@ const OnboardingEdit: FunctionComponent< BlockEditProps< Attributes > > = () =>
+				{ /*	TODO - Remove LoadingPlaceholder from Route and have it render via a store param.
+					Requires that we clear store after loading into editor. */ }
+				<Route exact path={ Step.SiteLoading }>
+					<LoadingPlaceholder />
```
```suggestion
					<CreateSite />
```

sirreal commented on client/landing/gutenboarding/onboarding-block/edit.tsx:
```diff
@@ -70,6 +71,11 @@ const OnboardingEdit: FunctionComponent< BlockEditProps< Attributes > > = () =>
 				</Route>
+				{ /*	TODO - Remove LoadingPlaceholder from Route and have it render via a store param.
+					Requires that we clear store after loading into editor. */ }
+				<Route exact path={ Step.SiteLoading }>
```
```suggestion
				<Route exact path={ Step.CreateSite }>
```

sirreal commented on client/landing/gutenboarding/onboarding-block/edit.tsx:
```diff
@@ -16,6 +16,7 @@ import StepperWizard from './stepper-wizard';
 import VerticalSelect from './vertical-select';
 import SignupForm from './signup-form';
 import SiteTitle from './site-title';
+import LoadingPlaceholder from './loading-placeholder';
```
```suggestion
import CreateSite from './create-site';
```

sirreal commented on client/landing/gutenboarding/onboarding-block/loading-placeholder/index.tsx:
```diff
@@ -0,0 +1,43 @@
+/**
```
This file would become `create-site`.

---

The following would be great in a follow up, but to align with a `create-site` route handling the intent, part of #39055 could be reverted and this part of the header would become a `<Link to={ Step.CreateSite } />`:

https://github.com/Automattic/wp-calypso/blob/2eff0c49dfee1fb27461bd1e11ee4e93d2d83e6e/client/landing/gutenboarding/components/header/index.tsx#L119-L128

Then this `<CreateSite />` component would be in charge of handling the side effect of site creation. We have a temporary implementation , but #39050 is adding the site store with `createSite` action we'd dispatch here.

For now, we can use the following, like the header was using:

https://github.com/Automattic/wp-calypso/blob/2eff0c49dfee1fb27461bd1e11ee4e93d2d83e6e/client/landing/gutenboarding/utils.ts#L27-L57

(☝️ we should return that `wpcom.undocumented().sitesNew( … )` promise and let this component handle navigation)

sirreal commented on client/landing/gutenboarding/onboarding-block/loading-placeholder/index.tsx:
```diff
@@ -0,0 +1,43 @@
+						</div>
+					</div>
+				</div>
+			</div>
```
I'd vote to get this merged with `Creating your site…` on a white background. We can update the loading view when we have clear design feedback. Does that seem OK @simison?

simison replied:
Yep! Sounds good. The goal is to just get something minimal in so that it's clear in the flow that something is loading. Doesn't matter too much how it looks like, as long as its minimal.

Addison-Stavlo replied:
Sounds good to me.  With that and with @simison 's suggestion of covering the entire UI, that will look something like this for now: 

![create-site-screen](https://user-images.githubusercontent.com/28742426/73866832-828bcd00-4813-11ea-9fed-595768a60aa4.gif)

After having the Spinner, it feels like something is missing between the last placeholder bar and the "Your site is being created" text.  But it will serve well enough as a temporary style until we something more concrete from design.

olesyabrk replied:
Hi, yes, can you add this as a needs-design ticket, if you haven't already? And tag the design team?

Addison-Stavlo replied:
@olesyabrk  sure thing! - I have created a ticket here: #39287

ockham commented on client/landing/gutenboarding/onboarding-block/loading-placeholder/index.tsx:
```diff
@@ -0,0 +1,43 @@
+export default function LoadingPlaceholder() {
+	return (
+		<div className="editor__placeholder">
+			<div className="edit-post-layout">
```
The `edit-post-*` `<div />`s in here seem to recreate Gutenberg's component hierarchy (for styling purposes, I assume). That feels kinda off _inside_ of a block. Is this needed in order to make the placeholder fullscreen-ish? Is there any way we can limit the class names in here to our block's scope?

(Going forward, I'd hope to make Gutenboarding completely `edit-post`-agnostic. After all, `edit-post` is the Gutenberg part that's specifically dedicated to post editing -- and here, we only want a slimmed down, headless editor.)

Addison-Stavlo replied:
> Is this needed in order to make the placeholder fullscreen-ish?

They aren't needed to make it fullscreen, but they are needed to put the blinking placeholder bars in their respective positions according to where they would appear in the standard editor.

> Is there any way we can limit the class names in here to our block's scope?

By this I assume you mean updating them from `.edit-post-header` to something like `.create-site__header` ?  If we did this the entire style and position of the placeholders will break down, requiring us to re-write all of the css that has already been defined for these classes.  This does not seem like a great use of time for a temporary loading style.

Once we have another editor for styles to mimic, these `.edit-post-*` classnames and placeholder structure could be updated to ensure the style is matching the new editor instead.  But that is assuming that design will decide to keep this as an editor loading screen and not some other 'creating your site' style design.

> Going forward, I'd hope to make Gutenboarding completely edit-post-agnostic. After all, edit-post is the Gutenberg part that's specifically dedicated to post editing

That definitely makes sense to me, but using `.edit-post-*` to piggy-back some pre-defined styles for the purpose of an isolated and temporary loading screen seems acceptable.

Addison-Stavlo replied:
Note - ignore the "outdated" flag, the directory was renamed to `create-site` which triggered that.

Addison-Stavlo replied:
I updated the new file under `create-site` to get rid of the `edit-post-*`s.

Addison-Stavlo commented:
* Updated some styles and naming conventions as per requests.
* Updated the header file to redirect to this when we start site creation.  (my apologies for saying the flow worked last night, I forgot to push changes to this file.)

I do still think we will want to update this to contain the create-site logic in a follow up as @sirreal suggests.  As well as make this Route redirect to another step if somehow loaded when site creation does not make sense.  But I think this is in a pretty good place to test and merge, with room for some followups after #39203 and  #39050 land.

![create-site-updated](https://user-images.githubusercontent.com/28742426/73876301-c509d580-4824-11ea-982b-429bcf4ca4ef.gif)

razvanpapadopol commented:
Rebased and added a redirect from `/create-site` to `/` when not creating a site.

Addison-Stavlo commented:
Awesome, everything looks good to me.  Thanks for rebasing and updating!

I think it looks good to ship.  With @sirreal's suggestion to update the Site creation button to a `<Link to={ Step.CreateSite } />` and move some of the creation logic into that component as a good follow-up once #39050 lands.
